### PR TITLE
Stateless scheduler

### PR DESCRIPTION
**Pull request by pcuenca**
This is the biggest change derived from #5.

In that branch I was previously using a standard Python dataclass and a `dict` to manage the scheduler state, and I converted between them when needed. I have now adopted `flax.struct.dataclass`, which is essentially the same but:
- It's frozen so it enforces functional updates.
- Can be easily replicated across devices.

I'm not super happy with the result for these reasons:
- Invoking `scheduler.set_timesteps()` requires more information to initialize the state, such as the shape of the expected inputs. This makes the code a bit uglier because you need to prepare that in advance.
- The dataclass has many state vars. In PyTorch we initialized the instance with just the betas and the number of training steps, and then added a few items when setting the inference timesteps. If we try to do the same here by providing defaults to most attributes, then Python synthesizes an initializer with just the `betas` and the `num_training_steps`, and then the `replace` method doesn't work. I had to create a specific class method to simplify the initial creation of the instance.
- The updates in `step` are harder to follow than the PyTorch logic.

Note also that the initial PRK steps have not been implemented yet (since we are skipping them by default). If we go with this approach they should be easy to add.

### CODE REVIEW

File: stable_diffusion_jax/scheduling_pndm.py [hunk 1]
```diff
@@ -196,25 +196,84 @@ def step_plms(
...
+            counter,
+            [counter_0, counter_1, counter_2, counter_3, counter_other],
+            state,
+        )
+
+        sample = state.sample
+        model_output = state.model_output
+        prev_sample = self._get_prev_sample(state, sample, timestep, prev_timestep, model_output)
 
         prev_sample = self._get_prev_sample(sample, timestep, prev_timestep, model_output)
```
Comment (patrickvonplaten) [inline stable_diffusion_jax/scheduling_pndm.py]:
```suggestion
```

this seems like it should be removed?

File: stable_diffusion_jax/scheduling_pndm.py [hunk 2]
```diff
@@ -196,25 +196,84 @@ def step_plms(
...
         self.counter += 1
```
Comment (patrickvonplaten) [inline stable_diffusion_jax/scheduling_pndm.py]: do we still use the `self.counter` variable?

Comment (patrickvonplaten) [reply stable_diffusion_jax/scheduling_pndm.py]: Also we should return the state somewhere no?

File: stable_diffusion_jax/scheduling_pndm.py [hunk 3]
```diff
@@ -196,25 +196,84 @@ def step_plms(
...
+                ets = ets,
+                model_output = next_model_output,
+                sample = sample,
+            )
+
+        counter = jnp.clip(state.counter, 0, 4)
+        state = jax.lax.switch(
+            counter,
+            [counter_0, counter_1, counter_2, counter_3, counter_other],
+            state,
```
Comment (patrickvonplaten) [inline stable_diffusion_jax/scheduling_pndm.py:289]: the state doesn't seem to be passed to the `step` method

**Comment by patrickvonplaten** [comment]
That's a great PR @pcuenca! Think some clean-ups are left to make it functional, but overall the design seems like the correct design to me

**Comment by pcuenca** [comment]
@patrickvonplaten You are right! I stashed my changes to apply them in a new branch and messed up somehow. I have just pushed the correct version, I hope. Sorry about that!

File: stable_diffusion_jax/scheduling_pndm.py (hunk 3 above)
Comment (pcuenca) [reply stable_diffusion_jax/scheduling_pndm.py:289]: It wasn't, I updated the code with the latest version.

File: stable_diffusion_jax/scheduling_pndm.py (hunk 1 above)
Comment (pcuenca) [reply stable_diffusion_jax/scheduling_pndm.py]: 👍

**Review by patrickvonplaten — APPROVED** [review]
Cool - looks clean to me now!
